distributed/deploy/spec.py

```diff
@@ -0,0 +1,296 @@
+                    await task
+            for name in to_close:
+                del self.workers[name]
+
+            to_open = set(self.worker_spec) - set(self.workers)
+            workers = []
+            for name in to_open:
+                d = self.worker_spec[name]
+                cls, opts = d["cls"], d.get("options", {})
+                if "name" not in opts:
+                    opts = toolz.merge({"name": name}, opts, {"loop": self.loop})
+                worker = cls(self.scheduler.address, **opts)
+                self._created.add(worker)
+                workers.append(worker)
+            if workers:
+                await asyncio.wait(workers)
+                for w in workers:
+                    w._cluster = weakref.ref(self)
+                    if self.status == "running":
+                        await w
+            self.workers.update(dict(zip(to_open, workers)))
+
+    def __await__(self):
+        async def _():
+            if self.status == "created":
+                await self._start()
+            await self.scheduler
+            await self._correct_state()
+            if self.workers:
+                await asyncio.wait(list(self.workers.values()))  # maybe there are more
+            await self._wait_for_workers()
+            return self
+
+        return _().__await__()
+
+    async def _wait_for_workers(self):
+        # TODO: this function needs to query scheduler and worker state
+        # remotely without assuming that they are local
+        while {d["name"] for d in self.scheduler.identity()["workers"].values()} != set(
+            self.workers
+        ):
+            if (
+                any(w.status == "closed" for w in self.workers.values())
+                and self.scheduler.status == "running"
+            ):
+                raise gen.TimeoutError("Worker unexpectedly closed")
+            await asyncio.sleep(0.1)
+
+    async def __aenter__(self):
+        await self
+        return self
+
+    async def __aexit__(self, typ, value, traceback):
+        await self.close()
+
+    async def _close(self):
+        while self.status == "closing":
+            await asyncio.sleep(0.1)
+        if self.status == "closed":
+            return
+        self.status = "closing"
+
+        async with self._lock:
+            await self.scheduler.close(close_workers=True)
+        self.scale(0)
+        await self._correct_state()
+        for w in self._created:
+            assert w.status == "closed"
+
+        if hasattr(self, "_old_logging_level"):
+            silence_logging(self._old_logging_level)
+
+        self.status = "closed"
+
+    def close(self):
+        with ignoring(RuntimeError):  # loop closed during process shutdown
+            return self.sync(self._close)
+
+    def __del__(self):
+        if self.status != "closed":
+            self.close()
+
+    def __enter__(self):
+        self.sync(self._correct_state)
+        self.sync(self._wait_for_workers)
```

Comment:
Does this mean that `__enter__` will only complete once the initial n workers have started? What happens if we request 2, 1 worker starts and 1 fails?

Reply:
Yes, this might hang.  I'm not sure we ever had a test in our test suite with this case.  I'll add something.

Reply:
Added a test in 5e940691